SCVMM Sprout additions and other optimizations

jteehan opened the pull request.

dajoRH commented:
Lint report for commit 56ff31f304e080063fb609d3577581a89d911416:
:godmode: All seems good! :cake: :punch: :cookie: 
*CFME QE Bot*

jteehan commented:
Here is a Unit Test File which you can use or modify to do method verifications.
http://pastebin.test.redhat.com/346060

The list_vm and new all_vms run three times quicker by only returning the property or properties we need.  The data returned in list_vm has been reduced by over 99%  45K vs 5.6Mb

You need the cfme_data.yaml which I checked in earlier.

Have Fun!

dajoRH commented:
Lint report for commit 1fe7d6479bc2e7fa87bf54ee6a893d8be3df1aa2:
:godmode: All seems good! :cake: :punch: :cookie: 
*CFME QE Bot*

mfalesni commented on mgmtsystem/scvmm.py:
```diff
@@ -174,31 +230,52 @@ def wait_vm_suspended(self, vm_name, num_sec=300):
+         """.format(vm_name=vm_name, ls=library, lp=library_share)
+        self.logger.info(" Creating SCVMM Template `{vm_name}` from VM `{vm_name}`-tpl"
+            .format(vm_name=vm_name))
+        self.run_script(script)
```
So, here what it should do is that it should rename the original VM (testvm) to eg (testvm_copytemplate), then it should create a template (testvm) from (testvm_copytemplate) and delete the (testvm_copytemplate) so in the end, only the template exists and has the same name as the original VM. Basically just two more calls wrapping this script call :)

jteehan replied:
Easy enough.  The VM clone works fine so I'll just integrate it.

jteehan replied:
I'll edit out the -tpl suffix on the template and that should do it.

dajoRH commented:
Lint report for commit 8962b2db01ad931bd18ffc904ed6da81314e2d52:
:godmode: All seems good! :cake: :punch: :cookie: 
*CFME QE Bot*

mfalesni commented on mgmtsystem/scvmm.py:
```diff
@@ -103,26 +103,82 @@ def wait_vm_stopped(self, vm_name, num_sec=300):
+        for vm in lines:
+            fields = vm.split()
+            vm_data = fields[0], fields[1], fields[2], fields[3]
+            vm_list.append(vm_data)
```
2 more things need to be done here:

At the top:
```python
from .base import VMInfo   # I think, or maybe mgmtsystem.base, it depends
```

And here:
```python
vm_data = fields[0], fields[1], fields[2], None if fields[3] == 'None' else fields[3]
vm_list.append(VMInfo(*vm_data))
```

I wrap the result in a namedtuple for easy access and also the None cannot be a string but a real None. And it definitely should not return a string, just the `vm_list`

jteehan replied:
Done

For 1, I just copied what was in RHEV

Didn't know about #2.  That looks much much better.  Thanks.

dajoRH commented:
Lint report for commit db67b1824359915c055181f5872b77aa32e342a5:
:godmode: All seems good! :cake: :punch: :cookie: 
*CFME QE Bot*

mfalesni commented on mgmtsystem/scvmm.py:
```diff
@@ -103,26 +103,82 @@ def wait_vm_stopped(self, vm_name, num_sec=300):
+            fields = vm.split()
+            vm_data = fields[0], fields[1], fields[2], None if fields[3] == 'None' else fields[3]
+            vm_list.append(VMInfo(*vm_data))
+        return str(vm_list)
```
Please, remove the string conversion as I requested in the comment before.

dajoRH commented:
Lint report for commit da2324fb6a5d5093e07f6eb6fbe6ef8579fe6258:
:godmode: All seems good! :cake: :punch: :cookie: 
*CFME QE Bot*

mfalesni commented:
Yeah, the `all_vms()` looks much better now :)

dajoRH commented:
Lint report for commit f700012c56784337a5c1c44c5fb779ee75d687b5:
:godmode: All seems good! :cake: :punch: :cookie: 
*CFME QE Bot*

dajoRH commented:
Lint report for commit 10579f83b46ba938e1143f6ba824d82d6627178e:
:godmode: All seems good! :cake: :punch: :cookie: 
*CFME QE Bot*

jteehan commented:
I think I got it all squashed up.  Anything else preventing this from being merged?  Happy to clean up anything.

mfalesni commented:
Lemme try it out and see how does it work.

dajoRH commented:
Lint report for commit c4c82a765014ea2b8a6ffbd27bc2da0950955840:
:godmode: All seems good! :cake: :punch: :cookie: 
*CFME QE Bot*

jteehan commented:
How do you want to handle VMs that return multiple IP Addresses?  Also, failures don't have a GUID, so I'm currently returning None for that.

mfalesni commented:
@jteehan Good point. Currently, mgmtsystem returns the one that is "available". That means if an openstack instance, for example, has the internal address and a floating IP, it returns the floating IP because that one is accessible to the machine you run on, you cannot reach the internal IP. In the case the VM has only the internally routed address, current_ip_address returns None for example and so does all_vms. So if something similar happens inside scvmm, then this is the model example. Also it returns IPV4 only so far. If there are multiple IP addresses that can be reached from the machine you are running on, then pick the one that is the least likely to change/be deleted. Sprout does not care about the address as long as it can be connected through, but still, it is better to have it stable :)

By failures you mean VMs that are in Failed state or something? If that is the case, it should not be a problem. I can filter Nones out in the sync task that assigns GUIDs based on names.

jteehan commented:
Thanks.  I tried to set up as many situations as I could think of to see what would happen.  We're trying to fix the multiple IP address with CFME as well.  All IPs in the list are valid, and I only return from SCVMM IPv4Addresses as you may see in the powershell code.  When there are two it returns a funky format that I'll have to parse manually.  Another line or two of code.

For the GUID = None, I saw a VM that failed to create, but seem to have gotten past the point of no return so that it is still in the database.  I'm trapping for that and returning None just as we did for the IP address.  It's rare, but it may happen and I wanted you to be aware of it.  Thanks again.  It's looking really good!  Lot's of error checking which I like.

dajoRH commented:
Lint report for commit 0f23da641bea83fd3b70dd5c058c5225120ecbfa:
:godmode: All seems good! :cake: :punch: :cookie: 
*CFME QE Bot*

jteehan commented:
Okay, so this VM has two valid IP addresses "ztest-26-018", and I'm returning just the first one.  I think I've covered everything on the list.

mfalesni commented:
I see, but I was wondering - it is not possible to use some format to transfer tha data rather tahn splitting and stuff inside python? That would make me a bit more comfortable.

jteehan commented:
I reran the script again and looked at the changes at closely as I am able to do.  Everything looks correct to the ever changing specification.  I replaced the imaginary PS cmdlets with real ones.  I rewrote the inefficient methods with methods that are orders of magnitude faster than the originals.  But I haven't clue how to make the module more "comfortable".  Please forward a link to the "comfortable" model, or provide me a specific implementation request.  Otherwise, approve the PR.

mfalesni commented:
@jteehan I meant instead of relying on separators and parsing whitespace, can you make PowerShell emit an XML that will get parsed using xpath on client side? From what I know from writing the original module, one can use `ConvertTo-Xml` to convert the objects to XML so if it is possible to slip this call somewhere in the end of the chain to produce an XML, that would be much better than "parsing" approach.

mfalesni commented:
To avoid possible escaping issue where a rogue character would screw the parsing up. :)

jteehan commented:
That's always a risk I suppose.  I'm just not seeing how that risk is reduced by converting the output TO xml just so that on the very next line I can convert it FROM xml.  ConvertTo-xml is just going to parse the white space like I did.  But, if it's that important to you, I'll get to it when I can.  Thanks.

mfalesni commented:
Well, if you can dump the array of $outobj through ConvertTo-XML, then it should be relatively straightforward to extract the data on the python side and we will be sure that it is what we want it to be :)

dajoRH commented:
Lint report for commit 61c23aa841fa80b80bef198a2dde2a3f679be460:

`mgmtsystem/scvmm.py`:
- :bangbang: Line 172:12: **E225** *missing whitespace around operator*
- :bangbang: Line 183:17: **E123** *closing bracket does not match indentation of opening bracket's line*

Please, rectify these issues :smirk: .
*CFME QE Bot*

dajoRH commented:
Lint report for commit dae5dc069b5b7ccb7998717f526f672090b8485a:

Everything seems all right :smile: .
*CFME QE Bot*

dajoRH commented:
Lint report for commit 1a7151f151034e24e5300b472bafa978a51a2b60:

Everything seems all right :smile: .
*CFME QE Bot*

mfalesni commented:
Okay, this seems good to go. I tried most of the functions and it appears that it works properly. @psav let's merge and release :)
